libs/qa/common_df_diff.py

```diff
@@ -0,0 +1,138 @@
+import numpy as np
+from pydantic import BaseModel
+
+from covidactnow.datapublic.common_fields import CommonFields, COMMON_FIELDS_TIMESERIES_KEYS
+
+
+class DatasetDiff(BaseModel):
+    duplicates_dropped: pd.DataFrame
+    melt: pd.DataFrame
+    all_variable_fips: pd.MultiIndex
+    my_ts: Optional[pd.MultiIndex] = None
+    common_fips: Optional[pd.DataFrame] = None
+    my_ts_points: Optional[pd.DataFrame] = None
+    ts_diffs: Optional[pd.DataFrame] = None
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def __str__(self):
+        return f"""Duplicate rows in this file: {self.duplicates_dropped}
+TS only in this file: {self.my_ts}
+TS points only in this file: {self.my_ts_points.groupby('date').size().to_dict()}
+TS diffs: {self.ts_diffs if self.ts_diffs is not None else ''}
+TS diffs: {self.ts_diffs.groupby('variable has_overlap'.split()).mean() if self.ts_diffs is not None else ''}
+"""
+
+    @staticmethod
+    def make(df: pd.DataFrame) -> "DatasetDiff":
+        dups = df.loc[df.index.duplicated(keep=False)]
+        if not dups.empty:
+            df = df.drop_duplicates()
+
+        df = df.reset_index().replace({pd.NA: np.nan}).convert_dtypes()
+        df[CommonFields.DATE] = pd.to_datetime(df[CommonFields.DATE])
+        # Drop columns that don't really contain timeseries values.
+        columns_to_drop = {
+            "index",
+            CommonFields.STATE,
+            CommonFields.COUNTRY,
+            CommonFields.AGGREGATE_LEVEL,
+        }.intersection(df.columns)
+        # Drop string columns because timeseries_diff can't handle them yet.
+        for col in df.select_dtypes(include={"object", "string"}):
+            if col != "fips":
+                print(f"Dropping field '{col}' based on type")
+                columns_to_drop.add(col)
+        if columns_to_drop:
+            df = df.drop(columns=columns_to_drop)
+
+        melt = (
+            df.melt(id_vars=COMMON_FIELDS_TIMESERIES_KEYS)
+            .set_index(["variable"] + COMMON_FIELDS_TIMESERIES_KEYS)
+            .dropna()
+        )
+        # When Int64 and float64 columns are merged into one by melt the 'object' dtype is used, which
+        # is not supported by timeseries_diff. Force 'value' back to a numeric dtype.
+        melt["value"] = pd.to_numeric(melt["value"])
+
+        all_variable_fips = melt.groupby("variable fips".split()).first().index
+        return DatasetDiff(duplicates_dropped=dups, melt=melt, all_variable_fips=all_variable_fips)
+
+    def compare(self, other: "DatasetDiff"):
+        # A timeseries is a set of <date, value> tuples identified by a <variable, fips> tuple. First
+        # find all the timeseries that have at least one real value in only one dataset. These are
+        # timeseries that appear in one file but not the other file.
+        self.my_ts = self.all_variable_fips.difference(other.all_variable_fips)
+        other.my_ts = other.all_variable_fips.difference(self.all_variable_fips)
```

Comment:
I'm having a bit of trouble with `my_ts`. I think I expect "ts" to be a pandas Series or DataFrame, so it's throwing me that it's a multi index? I don't think I quite understand this comment: "A timeseries is a set of <date, value> tuples identified by a <variable, fips> tuple"